crates/turborepo-ui/src/wui/server.rs

```diff
@@ -0,0 +1,106 @@
+use std::{cell::RefCell, sync::Arc};
+
+use async_graphql::{
+    http::GraphiQLSource, EmptyMutation, EmptySubscription, Object, Schema, SimpleObject,
+};
+use async_graphql_axum::GraphQL;
+use axum::{http::Method, response, response::IntoResponse, routing::get, Router};
+use serde::Serialize;
+use tokio::{net::TcpListener, sync::Mutex};
+use tower_http::cors::{Any, CorsLayer};
+
+use crate::wui::{
+    event::WebUIEvent,
+    subscriber::{Subscriber, TaskState, WebUIState},
+    Error,
+};
+
+#[derive(Debug, Clone, Serialize, SimpleObject)]
+struct Task {
+    name: String,
+    state: TaskState,
+}
+
+struct CurrentRun<'a> {
+    state: &'a Arc<Mutex<RefCell<WebUIState>>>,
+}
+
+#[Object]
+impl<'a> CurrentRun<'a> {
+    async fn tasks(&self) -> Vec<Task> {
+        self.state
+            .lock()
+            .await
+            .borrow()
+            .tasks()
+            .iter()
+            .map(|(task, state)| Task {
+                name: task.clone(),
+                state: state.clone(),
+            })
+            .collect()
+    }
+}
+
+pub struct Query {
+    state: Arc<Mutex<RefCell<WebUIState>>>,
+}
+
+impl Query {
+    pub fn new(state: Arc<Mutex<RefCell<WebUIState>>>) -> Self {
+        Self { state }
+    }
+}
+
+#[Object]
+impl Query {
+    async fn current_run(&self) -> CurrentRun {
+        CurrentRun { state: &self.state }
+    }
+}
+
+async fn graphiql() -> impl IntoResponse {
+    response::Html(
+        GraphiQLSource::build()
+            .endpoint("/")
+            .subscription_endpoint("/subscriptions")
+            .finish(),
+    )
+}
+
+pub async fn start_server(
+    rx: tokio::sync::mpsc::UnboundedReceiver<WebUIEvent>,
+) -> Result<(), crate::Error> {
+    let state = Arc::new(Mutex::new(RefCell::new(WebUIState::default())));
```

Comment:
I'm not sure if I understand the use of `RefCell` inside of the `Mutex`. Doesn't the `Mutex` already ensure that there's only ever one access to the underlying data at a time? Why are dynamic borrows necessary inside of the restriction?

If this is necessary a quick comment explaining why would be very appreciated.